### PR TITLE
Added a file filter based on regular expressions

### PR DESCRIPTION
**Pull request by mried**
The file filter is included within the ihate plugin and uses the new `import_task_created` event which is fired for each task after it is created.
The plugin is only a first approach. Maybe the plugin rules are much to complicated. I'm not even sure if I'll use them for my own library...
A new pipeline stage is used to fire the event so the code to be changed is limited to one place.
<hr/>
Note: Running `tox -e py27cov` or even `nose` on my windows box lets fail all the new tests of `test_ihate`:
```
beets: WARNING: ** error loading plugin ihate
beets: WARNING: Traceback (most recent call last):
  File "D:\work\beets\beets\beets\plugins.py", line 211, in load_plugins
    for obj in getattr(namespace, name).__dict__.values():
AttributeError: 'module' object has no attribute 'ihate'
```
Running the test module without the other modules is working. I found out the the paths in `beetsplug.__path__` are different but couldn't figure out what's leading to the difference nor why one of values leads to failures. Seems I'm at the end of my python knowledge :wink:...

### CODE REVIEW

**Comment by sampsyo** [comment]
Interesting issues. Thank you so much for being thorough in checking out all the options. You've definitely done this well, but you're absolutely right that this approach is a little limiting (and we could still prevent some use cases—which I think is also a problem with the old pipeline-stage version too).

I agree that we should just add the event if we can't find something perfect, but it would be even cooler to find one consistent interface that works for everything here. Here are two final ideas to consider:

* Do some refactoring to make sure all tasks are produced in the ImportTaskFactory. This would require some cleanup to centralize the logic. Then we provide an event that's only emitted from `ImportTaskFactory.tasks`. This would also require that the event handler be able to *return* new tasks, also, which is something events can't currently do but something I'm interested in adding anyway. (I'll help with that if we choose this option.)
* Add an alternative to the current plugin-provided pipeline stages that adds new stages at this point in the process. `early_import_stage` or something? We'd then have to resolve the other issues above: making sure these appear in the mini-pipelines, etc.

I think I lean toward the first option, even though it will require some nearer-term refactoring before we can get it working. Any thoughts?

**Comment by mried** [comment]
I played around a bit and tried to imitate the behaviour of `group_albums` which does the same thing I want do do (add new tasks to the pipeline). The plugin is uploaded here (beside other plugins :smile:): https://github.com/mried/beetsplug/blob/autosingleton/beetsplug/autosingleton.py

The idea is simple: If a false album task is detected, the action of the album task is set to skip. New `SingletonImportTask`s are emitted to the pipeline using `pipeline.multiple(...)`. But that seems not to work at all :disappointed:.

But maybe this is a third idea: Provide a way for plugins to add more tasks to the pipeline. I'm completely unsure if this is possible since I didn't understand the pipeline code right now :confused:.

Anyway: Your second idea came into my mind, too. But I thought about a more general solution: A pipeline stage should know its dependent stages (*not before* and *not after*). This way, the pipeline will be set up dynamically.

The first solution is a very flexible approach, hence with a lot of work :wink:. It would be much better if the `ImportTaskFactory` is responsible for the creation of *all* `ImportTask`s. But I don't see how `tasks` can provide all needed tasks. Right now it's a method to return all needed tasks for an import of files. What about the tasks created elsewhere, like in `group_albums` or `user_query`? Maybe I just didn't understand you right...

BTW: Happy new year! :tada:

**Comment by sampsyo** [comment]
Hmm… I'm not sure why `pipeline.multiple` wouldn't work for you. We use that a few times in the importer—any additional detail about what exactly happens instead?

And no, you're quite right that this much centralization might be too hard to achieve. But even so, something that reduces the number of places that we construct tasks would help with the overall agenda of letting task construction be customizable. I don't see a perfect solution at the moment, so perhaps some creativity that has yet to dawn on us will be helpful…

**Comment by mried** [comment]
Ah! I have to feed the result of `pipeline.multiple` to the pipeline afterwards. I thought this is a class method doing so for me :disappointed:. So what happened when using my plugin? Exactly nothing :smile:.

I'll try to move all the task generation into the `ImportTaskFactory` and prepare a mechanism to change them by a plugin event.

**Comment by mried** [comment]
OK, I did something to get these automatic singleton detection working. And it seems to work :tada:!

To do so, I had to change `plugins.send` and I'm completely unsure if I did it the right way. Right now, every return values of every plugin is added to a list and this list is returned afterwards.

Another big change had to be done in `importer.py` because now it is possible at all times that a task gets splitted into several tasks by a plugin, regardless of the type of task. Maybe that's not what we really want.

Any thoughts?

File: beets/plugins.py [hunk 1]
```diff
@@ -399,11 +399,18 @@ def send(event, **arguments):
...
     """
     log.debug(u'Sending event: {0}'.format(event))
+    return_values = []
     for handler in event_handlers()[event]:
         # Don't break legacy plugins if we want to pass more arguments
         argspec = inspect.getargspec(handler).args
         args = dict((k, v) for k, v in arguments.items() if k in argspec)
-        handler(**args)
+        result = handler(**args)
+        if isinstance(result, list):
```
Comment (sampsyo) [inline beets/plugins.py]: This special list handling could bite us when other sequence types are involved, or when separate lists are actually significant. Let's just append all the responses to the return list (although we should probably filter out `None`s for convenience) and let clients join lists if they want that.

**Comment by sampsyo** [comment]
Great! That modification to `plugins.send` was less painful than I imagined it would be. Let's get that into place, with documentation and everything, and make sure we're happy with it, and then return to finishing this plugin merge.

I share your worry about this:

> Another big change had to be done in importer.py because now it is possible at all times that a task gets splitted into several tasks by a plugin, regardless of the type of task. Maybe that's not what we really want.

It would be great to abstract this away. I'll continue thinking about how limit this effect while adding the event stuff.

I know I've said this a lot, but *thank you* for your patience here. I'm not ordinarily this pedantic and slow about design decisions, but the importer workflow is already incredibly complex, so it's a place where I'm especially sensitive about adding even small changes to the pipeline. We'll get there soon!

**Comment by mried** [comment]
Here we go :laughing: 

I was able to reduce the changes to the old importer code with moving some checks for `None` to a more central place. (These lead to the broken tests before...)

All tasks (including archive and sentinel) are created using the task factory and create the event.

Documentation and test are also added.

You don't have to excuse yourself for being pedantic. I would be as pedantic as you here :wink:. It's always better to think a bit about a change to find a good design rather implementing just *something* which works but is not understandable or won't work in all cases...

**Comment by mried** [comment]
I've merged the recent changes and it still works :smiley:. What else needs to be done on this topic?

**Comment by sampsyo** [comment]
Awesome; thank you for updating. I'm sorry for disappearing recently; beets has been really active lately and I've been low on time. I'm going to oscillate back to this as soon as I can—I think this is very nearly good to go if not ready already.

File: beets/importer.py [hunk 2]
```diff
@@ -1085,7 +1122,7 @@ def read_tasks(session):
 
             log.debug(u'extracting archive {0}',
                       displayable_path(toppath))
-            archive_task = ArchiveImportTask(toppath)
+            archive_task = task_factory.archive(toppath)[0]
             try:
                 archive_task.extract()
```
Comment (sampsyo) [inline beets/importer.py]: This is now incorrect, right? `archive_task` here is at least a list, and may not even be a list of `ArchiveTask`s if a plugin produces something else.

**Comment by sampsyo** [comment]
I'm starting now by merging the refactoring pieces first. (I'm hoping that this will make each piece of the puzzle easier to reason about.) I plan to open a new PR that isolates the plugin hook piece next.

File: beets/importer.py (hunk 2 above)
Comment (mried) [inline beets/importer.py]: Yes, you're right.
But fixing this is a bit more complicated: We can execute every archive task, that's easy. But which path will we use for further importing? Two archive tasks may place the extracted files on different locations.
I'm not sure how to solve this. We could only allow a single archive task and stop (or continue with the next file) when more than one archive task is returned. We could also only execute the first or last archive task. Or execute every task but only use the path of the last task.

**Comment by mried** [comment]
I've implemented the "execute all archive tasks and only use the path of the last task" solution but I'm still not sure if this is a good idea... Let me know what you think.

**Comment by sampsyo** [comment]
Alright—I've landed a compromise solution in dc5a79e.

That piece adds the `import_task_created` event in a much simpler way. It's more limited, but it should be sufficient for this new feature. There's no more useful return value from the event—since you've opted for using the "set skip" option anyway, I don't think we need it. This makes the modifications to `importer.py` much smaller—there are only four functions where tasks are created, so we just need calls in four places:

* `read_tasks`
* `query_tasks`
* `user_query`'s "as tracks" generator
* `group_albums`

Does this all look OK to you? If so, I'll get started on porting the rest of the PR over.

**Comment by mried** [comment]
Beside commit b7125e434315b00584cb6d6abf8baf5d26fb3bf0 (see my comment there), everything looks fine for the simple filter (which is all I added to the `ihate`plugin). Plugins are able to skip a task.

Can I help somehow on this task?

**Comment by sampsyo** [comment]
Great! I'm glad this seems to be what we need.

I'd be happy to have your help if you're not burned out on contributing to this feature. :smiley: As capricious and unnecessary as it seems, I've been thinking more about this as I reviewed it further and I'm no longer so sure that merging with `ihate` was such a brilliant idea in the first place. In retrospect, the purposes are substantially different—and it might be better, long-term, for your new functionality to live in its own small, focused plugin. How would you feel about going back on that decision? Would that be too frustrating of a reversal?

**Comment by mried** [comment]
> Great! I'm glad this seems to be what we need.

As long as we go on and add the rest of the PR :wink:

The filter stuff is placed at a separate plugin like it was at the beginning of the PR. And: no, it's not frustrating as long as we are not in a loop :smiley:

**Comment by sampsyo** [comment]
Great! I was thinking and now I believe that we should merge it with `ihate` again JUST KIDDING SORRY

Thank you for sticking with this for so long. I'll merge this now. :tada:

**Comment by sampsyo** [comment]
Okay, all merged up!

I hope this wasn't too presumptive of me, but I renamed the plugin to just `filefilter`. I think this makes the purpose a little bit clearer.

Thank you again for all your effort! You're a true hero. :sparkles:

**Comment by mried** [comment]
Great! :+1: 

But I guess I got you wrong some comments ago. I thought you wanted to add the "task replacement by plugins feature" as well to make it possible to implement #1167. Is this still an option?

**Comment by sampsyo** [comment]
No, you read me right—I just wanted to take this one step at a time. We can figure out how to evolve this minimally for the other issue.

**Pull request merged**